components/raftstore/src/store/peer.rs

```diff
@@ -2655,26 +2651,51 @@ where
         }
     }
 
+    /// Cancel the snapshot generation task under the following conditions:
+    ///
+    /// 1. The requesting peer is removed by a configuration change.
+    /// 2. The requesting peer becomes unreachable, e.g., TiKV goes down or a
+    ///    network partition occurs.
+    pub fn maybe_cancel_gen_snap_task(&self, unreachable_store_id: Option<u64>) {
+        let to_peer = match self.get_store().get_gen_snap_task().as_ref() {
+            Some(task) => task.to_peer.clone(),
+            None => return,
+        };
+        let cancel_by_unreachable_store =
+            unreachable_store_id.map_or(false, |s| to_peer.get_store_id() == s);
+        let to_peer = to_peer.clone();
+        let cancel_by_peer_not_found = find_peer_by_id(self.region(), to_peer.get_id()).is_none();
+        if cancel_by_unreachable_store || cancel_by_peer_not_found {
```

Comment:
The two canceling scenarios here look reasonable to me. I might be overthinking but if we miss some other scenarios that should be considered then we might run into the same issue again.

Not strictly related to this PR, do we consider having an online config option to turn off the precheck mechanism? That might be an overkill though.